qa: add tests for immutable object cache

CongMinYin opened the pull request:
based on qemu task, use immutable_object_cache task to test cache
based on rbd_fio task, use immutable_object_cache task to test cache

Signed-off-by: Yin Congmin <congmin.yin@intel.com>
Signed-off-by: Feng Hualong <hualong.feng@intel.com>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

CongMinYin commented:
Hi, jason  @dillaman . I am Congmin, Dehao doesn't do this work anymore, I'll take over it.
I modified the code according to your comment.  I have simplified the code structure. Immutable_object_cache.py only does its own tasks(daemon).  Thanks.

dillaman commented on qa/suites/rbd/ro/clusters/fix-3.yaml:
```diff
@@ -0,0 +1,8 @@
+roles:
+- [mon.a, mon.c, mgr.x, osd.0, osd.1, osd.2, osd.3]
```
Nit: you can probably drop this down to a two node test and reduce the number of MONs and OSDs. The fewer the number of required nodes, the easier it is to schedule the test.

dillaman commented on qa/suites/rbd/ro/clusters/fix-3.yaml:
```diff
@@ -0,0 +1,8 @@
+roles:
```
Nit: instead of `ro` as the test subdirectory, perhaps just `parent_cache`?

dillaman commented on qa/suites/rbd/ro/workloads/immutable_object_cache.yaml:
```diff
@@ -0,0 +1,24 @@
+tasks:
+- install:
```
Nit: pull all the common config / test setup to a single yaml

CongMinYin replied:
merge qemu test and fio test to a single yaml?

CongMinYin replied:
> Nit: pull all the common config / test setup to a single yaml

merge qemu test and fio test to a single yaml?

dillaman replied:
No, avoid duplicating these bootstrapping actions by pulling the common bits out of the two workload YAMLs and into its own YAML file in its own sub-directory.

dillaman commented on qa/tasks/immutable_object_cache.py:
```diff
@@ -0,0 +1,143 @@
+@contextlib.contextmanager
+def task(ctx, config):
+    """
+    basedun_qemu(ctx=ctx, config=conf on implement of qemu task.
```
Nit: lots of `qemu` references

CongMinYin replied:
> Nit: lots of `qemu` references

sorry,  these are outdated information, I will delete.

dillaman commented on qa/tasks/immutable_object_cache.py:
```diff
@@ -0,0 +1,143 @@
+        log.info("cleanup thrashes immutable object cache")
+
+@contextlib.contextmanager
+def task(ctx, config):
```
Nit: I'd prefer to see this as two separate task files and avoid the `execution` command

CongMinYin replied:
Do you mean that parent cache setup and  thrashes test as two ".py" file?

CongMinYin replied:
> Nit: I'd prefer to see this as two separate task files and avoid the `execution` command

Do you mean that parent cache setup and thrashes test as two ".py" file?

dillaman replied:
Yes, two separate tasks like `rbd_mirror` and `rbd_mirror_thrash`.

CongMinYin commented:
@dillaman I changed the file name as you said,  split the cluster yaml file and task python file, pulled the common bits (ceph installation and immutable object cache start) out of the two workload YAMLs and put it into "pool" sub-directory.  Thanks.

dillaman commented on qa/suites/rbd/immutable-object-cache/workloads/qemu_on_immutable_object_cache_and_thrashes.yaml:
```diff
@@ -0,0 +1,11 @@
+      cpus: 4
+      memory: 4096
+      disks: 3
+- immutable_object_cache_thrashes:
```
Nit: 
```suggestion
- immutable_object_cache_thrash:
```

CongMinYin commented:
@dillaman changed "all" to "client.0", changed "thrashes" to "thrash" when it is as a noun. Thanks.

dillaman commented:
Looks like the test case task fails during cleanup:

http://pulpito.ceph.com/jdillaman-2020-05-29_11:50:19-rbd-wip-jd-testing-2-distro-basic-smithi/5103147/

CongMinYin commented:
> Looks like the test case task fails during cleanup:
> 
> http://pulpito.ceph.com/jdillaman-2020-05-29_11:50:19-rbd-wip-jd-testing-2-distro-basic-smithi/5103147/

Hi @dillaman , The failure occurs inside the qemu. I had this problem before because of a network problem (proxy). So inside the qemu, the task can't run ".sh" or "wget" correctly. I think it is not read only cache problem. Could you please run original qemu task to check? (qa/suits/rbd/qemu/)

On the other hand, it failed in  qemu "test" task, is not necessary foring test read only cache. Changing qa/run_xfstests_qemu.sh to true.sh can skip the unnecessary qemu task.  How do you think?

tasks:
- qemu:
    client.0:
      clone: true
      test: qa/run_xfstests_qemu.sh  ====> qa/workunits/true.sh

dillaman commented:
> On the other hand, it failed in qemu "test" task, is not necessary for testing read only cache. Changing qa/run_xfstests_qemu.sh to true.sh can skip the unnecessary qemu task. How do you think?

Well, essentially the QEMU test case failed to run [1] so the test failed. So you need to get it to work since otherwise it's not actually testing anything.

[1] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-05-29_11:50:19-rbd-wip-jd-testing-2-distro-basic-smithi/5103147/remote/smithi174/qemu/client.0/test.log

CongMinYin commented:
Hi @dillaman , retest please. 
Test case faild because that during thrash task, cache daemon was dead and  qemu read cache to start in the meanwhile.
The original scheme is to set a waiting time for QEMU to start, and then exec thrash task. But the QEMU startup time is unpredictable, so I put the thrash task before QEMU task, and deleted the code waiting for QEMU to start.
Thanks.

CongMinYin commented:
> Hi @dillaman , retest please.

dillaman commented:
>  Test case faild because that during thrash task, cache daemon was dead and qemu read cache to start in the meanwhile.

Hmm, the purpose of the thrashing is to prove that client IOs will continue uninterrupted should the daemon not be working properly. If QEMU really failed before the daemon wasn't running, maybe we should fix that. i.e. is it because we pass back errors on initialization? [1]

[1] https://github.com/ceph/ceph/blob/master/src/librbd/cache/ParentCacheObjectDispatch.cc#L187

CongMinYin commented:
> > Test case faild because that during thrash task, cache daemon was dead and qemu read cache to start in the meanwhile.
> 
> Hmm, the purpose of the thrashing is to prove that client IOs will continue uninterrupted should the daemon not be working properly. If QEMU really failed before the daemon was running, maybe we should fix that. i.e. is it because we pass back errors on initialization? [1]
> 
> [1] https://github.com/ceph/ceph/blob/master/src/librbd/cache/ParentCacheObjectDispatch.cc#L187

Yes, When We init parent cache, it  will create_cache_session, connect failed to RO daemon, and on_finish with ret < 0 if daemon is not runing but enabled. I think this is an inappropriate design.  Should I fix this first?

dillaman commented:
>  Should I fix this first?

Yes, please -- but you can just add another commit for fixing that to this PR (instead of opening another PR) since the teuthology test will help verify that it's fixed.

CongMinYin commented:
Hi @dillaman ,  I push another commit to fix reading child image failed bug when enabling immutable cache but daemon down. And "rebase" has been done. I tested, child image could be read  when daemon down. Then I did qa test with fixed ceph version and passed on my teuthology. Please do qa test in your teuthology using this fixed ceph version. 

Another issue: when daemon is down,  reading child image will output some "lderr" information, do you think it is necessary？
root@ceph05liang:~/cmyin/ceph/build# rbd info test/clone1
2020-06-10T02:13:05.376-0400 7f585ffff700 20 ceph::cache::CacheClient: 0x7f5860001500 handle_connect: fails to connect to cache server. error : Connection refused
2020-06-10T02:13:05.376-0400 7f585ffff700 20 ceph::cache::CacheClient: 0x7f5860001500 fault: fault.Connection refused
2020-06-10T02:13:05.376-0400 7f585ffff700 20 ceph::cache::CacheClient: 0x7f5860001500 fault: Connecting RO daenmon fails : Connection refused. Immutable-object-cache daemon is down ? Data will be read from ceph cluster 
2020-06-10T02:13:05.376-0400 7f585ffff700 -1 librbd::cache::ParentCacheObjectDispatch: 0x7f5860001430 operator(): Parent cache fail to connect RO daeomn.
2020-06-10T02:13:05.376-0400 7f585ffff700 -1 librbd::cache::ParentCacheObjectDispatch: 0x7f5860001430 operator(): Parent cache fail to register client.
2020-06-10T02:13:05.376-0400 7f585ffff700 -1 librbd::cache::ParentCacheObjectDispatch: 0x7f5860001430 handle_register_client: Parent cache register fails.
2020-06-10T02:13:05.376-0400 7f585ffff700 -1 librbd::image::0x7f5860001430 handle_init_plugin_registry: OpenRequest::failed to initialize plugin registry:.

CongMinYin commented:
Hi @dillaman ,  I push another commit to fix reading child image failed bug when enabling immutable cache but daemon down. After rebase, I also modify the "yaml" configure to enable parent cache as a plugin.  I test it can read child image when daemon down. Then I did qa test on my teuthology and passed. Please do qa test in your teuthology using this fixed ceph version. 

Another issue: when daemon is down,  reading child image will output some "lderr" information including parent cache part and plugin part, do you think it is necessary？
root@ceph05liang:~/cmyin/ceph/build# rbd info test/clone1
2020-06-10T02:13:05.376-0400 7f585ffff700 20 ceph::cache::CacheClient: 0x7f5860001500 handle_connect: fails to connect to cache server. error : Connection refused
2020-06-10T02:13:05.376-0400 7f585ffff700 20 ceph::cache::CacheClient: 0x7f5860001500 fault: fault.Connection refused
2020-06-10T02:13:05.376-0400 7f585ffff700 20 ceph::cache::CacheClient: 0x7f5860001500 fault: Connecting RO daenmon fails : Connection refused. Immutable-object-cache daemon is down ? Data will be read from ceph cluster 
2020-06-10T02:13:05.376-0400 7f585ffff700 -1 librbd::cache::ParentCacheObjectDispatch: 0x7f5860001430 operator(): Parent cache fail to connect RO daeomn.
2020-06-10T02:13:05.376-0400 7f585ffff700 -1 librbd::cache::ParentCacheObjectDispatch: 0x7f5860001430 operator(): Parent cache fail to register client.
2020-06-10T02:13:05.376-0400 7f585ffff700 -1 librbd::cache::ParentCacheObjectDispatch: 0x7f5860001430 handle_register_client: Parent cache register fails.
2020-06-10T02:13:05.376-0400 7f585ffff700 -1 librbd::image::0x7f5860001430 handle_init_plugin_registry: OpenRequest::failed to initialize plugin registry:.

CongMinYin commented:
Hi @dillaman , retest please.

dillaman commented:
Looks like the recent ASIO changes broke the plugin. I will take a look. However, it exposes an issue where the test doesn't fail even if the cache isn't running [1] and the QEMU test still looks like it had issues [2][3]

[1] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-06-17_21:02:35-rbd-wip-jd-testing-distro-basic-smithi/5158907/teuthology.log
[2] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-06-17_21:02:35-rbd-wip-jd-testing-distro-basic-smithi/5158908/teuthology.log
[3] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-06-17_21:02:35-rbd-wip-jd-testing-distro-basic-smithi/5158908/remote/smithi062/qemu/client.0/test.log

CongMinYin commented:
> Looks like the recent ASIO changes broke the plugin. I will take a look. However, it exposes an issue where the test doesn't fail even if the cache isn't running [1] and the QEMU test still looks like it had issues [2][3]
> 
> [1] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-06-17_21:02:35-rbd-wip-jd-testing-distro-basic-smithi/5158907/teuthology.log
> [2] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-06-17_21:02:35-rbd-wip-jd-testing-distro-basic-smithi/5158908/teuthology.log
> [3] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-06-17_21:02:35-rbd-wip-jd-testing-distro-basic-smithi/5158908/remote/smithi062/qemu/client.0/test.log

Yes, now, if the daemon dead, qemu also can read data from ceph, if the daemon run, read data from immutable object cache. Before the commit "fix bug read faild when daemon is down", qemu will read failed when daemon down. I think the former is we want.

And  my teuthology shows connected fail but exited 0 when running the script in qemu. Because our company proxy block(only http proxy, no git proxy). So I can do little about the internal tasks of QEMU.  Thank you for helping with QEMU issue.

+ '[' -n '' ']'
+ export TESTDIR=/tmp/cephtest
+ TESTDIR=/tmp/cephtest
+ '[' -d /tmp/cephtest ']'
+ mkdir /tmp/cephtest
+ URL_BASE='https://git.ceph.com/?p=ceph.git;a=blob_plain;f=qa'
+ SCRIPT=run_xfstests-obsolete.sh
+ cd /tmp/cephtest
+ wget -O run_xfstests-obsolete.sh 'https://git.ceph.com/?p=ceph.git;a=blob_plain;f=qa/run_xfstests-obsolete.sh'
--2020-06-12 07:41:39--  https://git.ceph.com/?p=ceph.git;a=blob_plain;f=qa/run_xfstests-obsolete.sh
Resolving git.ceph.com (git.ceph.com)... 8.43.84.132
Connecting to git.ceph.com (git.ceph.com)|8.43.84.132|:443... failed: Connection refused.
+ chmod +x run_xfstests-obsolete.sh
+ ./run_xfstests-obsolete.sh -c 1 -f xfs -t /dev/vdb -s /dev/vdc 1-7 9-17 19-26 28-49 51-61 63 66-67 69-79 83 85-105 108-110 112-135 137-170 174-191 193-204 206-217 220-227 230-231 233 235-241 243-249 251-262 264-278 281-286 288-289
+ STATUS=0
+ rm -f run_xfstests-obsolete.sh
+ exit 0

CongMinYin commented:
Hi @dillaman ,  the test log shows "./run_xfstests-obsolete.sh: line 288: make: command not found". Should “make” be installed in the “/run_xfstests-obsolete.sh”script first.

CongMinYin commented:
Hi @dillaman , this failure is caused by ubuntu image. I delete the url, the test can run success.  Can you retest it?
Another question, should I add "exclude_arch: armv7l" in the qemu task yam?(I saw it in qemu task )

CongMinYin commented:
jenkins retest this please

dillaman commented:
jenkins test dashboard backend

dillaman commented:
jenkins test dashboard backend

dillaman commented:
jenkins test dashboard backend
